widget/tree.go

```diff
@@ -0,0 +1,791 @@
+			m = r.treeContent.tree.branchMinSize
+		}
+		m.Width += depth * (theme.IconInlineSize() + theme.Padding())
+		min.Width = fyne.Max(min.Width, m.Width)
+		min.Height += m.Height
+	})
+	return
+}
+
+func (r *treeContentRenderer) Objects() (objects []fyne.CanvasObject) {
+	r.treeContent.propertyLock.RLock()
+	for _, d := range r.dividers {
+		objects = append(objects, d)
+	}
+	for _, b := range r.branches {
+		objects = append(objects, b)
+	}
+	for _, l := range r.leaves {
+		objects = append(objects, l)
+	}
+	r.treeContent.propertyLock.RUnlock()
+	return
+}
+
+func (r *treeContentRenderer) Refresh() {
+	s := r.treeContent.Size()
+	if s.IsZero() {
+		r.treeContent.Resize(r.treeContent.MinSize().Max(r.treeContent.tree.Size()))
+	} else {
+		r.Layout(s)
+	}
+	r.treeContent.propertyLock.RLock()
+	for _, d := range r.dividers {
+		d.FillColor = theme.ShadowColor()
+	}
+	for _, b := range r.branches {
+		b.Refresh()
+	}
+	for _, l := range r.leaves {
+		l.Refresh()
+	}
+	r.treeContent.propertyLock.RUnlock()
+	canvas.Refresh(r.treeContent.super())
+}
+
+func (r *treeContentRenderer) getBranch() (b *branch) {
+	o := r.branchPool.Obtain()
+	if o != nil {
+		b = o.(*branch)
+	} else {
+		var content fyne.CanvasObject
+		if f := r.treeContent.tree.CreateNode; f != nil {
+			content = f(true)
+		}
+		b = newBranch(r.treeContent.tree, content)
+	}
+	return
+}
+
+func (r *treeContentRenderer) getLeaf() (l *leaf) {
+	o := r.leafPool.Obtain()
+	if o != nil {
+		l = o.(*leaf)
+	} else {
+		var content fyne.CanvasObject
+		if f := r.treeContent.tree.CreateNode; f != nil {
+			content = f(false)
+		}
+		l = newLeaf(r.treeContent.tree, content)
+	}
+	return
+}
+
+var _ desktop.Hoverable = (*treeNode)(nil)
+var _ fyne.CanvasObject = (*treeNode)(nil)
+var _ fyne.Tappable = (*treeNode)(nil)
+
+type treeNode struct {
+	BaseWidget
+	tree    *Tree
+	uid     string
+	depth   int
+	hovered bool
+	icon    fyne.CanvasObject
+	content fyne.CanvasObject
+}
+
+func (n *treeNode) Content() fyne.CanvasObject {
+	return n.content
+}
+
+func (n *treeNode) CreateRenderer() fyne.WidgetRenderer {
+	return &treeNodeRenderer{
+		BaseRenderer: widget.BaseRenderer{},
+		treeNode:     n,
+		indicator:    canvas.NewRectangle(theme.BackgroundColor()),
+	}
+}
+
+func (n *treeNode) Indent() int {
+	return n.depth * (theme.IconInlineSize() + theme.Padding())
+}
+
+// MouseIn is called when a desktop pointer enters the widget
+func (n *treeNode) MouseIn(*desktop.MouseEvent) {
+	n.hovered = true
+	n.partialRefresh()
+}
+
+// MouseMoved is called when a desktop pointer hovers over the widget
+func (n *treeNode) MouseMoved(*desktop.MouseEvent) {
+}
+
+// MouseOut is called when a desktop pointer exits the widget
+func (n *treeNode) MouseOut() {
+	n.hovered = false
+	n.partialRefresh()
+}
+
+func (n *treeNode) Tapped(*fyne.PointEvent) {
+	n.tree.SetSelectedNode(n.uid)
+	if f := n.tree.OnNodeSelected; f != nil {
+		f(n.uid)
+	}
+}
+
+func (n *treeNode) Update(uid string, depth int) {
```

Comment:
This plus another few internal types have an exported Upate function, I'm not sure it should be - the exported methods above are satisfying an interface.